ffi/src/input.rs

```diff
@@ -1,11 +1,131 @@
+        }
+
+        pub fn as_operation(&self) -> Box<Operation> {
+            Box::new(Operation(ironrdp::input::Operation::MouseMove(self.0)))
+        }
+    }
+
+    #[diplomat::opaque]
+    pub struct MouseButton(pub ironrdp::input::MouseButton);
+
+    #[diplomat::enum_convert(ironrdp::input::MouseButton)]
+    pub enum MouseButtonType {
+        Left = 0,
+        Middle = 1,
+        Right = 2,
+        X1 = 3,
+        X2 = 4,
+    }
+
+    impl MouseButton {
+        pub fn new(button: MouseButtonType) -> Box<MouseButton> {
+            Box::new(MouseButton(button.into()))
+        }
+
+        pub fn as_operation_mouse_button_pressed(&self) -> Box<Operation> {
+            let operation = ironrdp::input::Operation::MouseButtonPressed(self.0);
+            Box::new(Operation(operation))
+        }
+
+        pub fn as_operation_mouse_button_released(&self) -> Box<Operation> {
+            let operation = ironrdp::input::Operation::MouseButtonReleased(self.0);
+            Box::new(Operation(operation))
+        }
+    }
+
+    #[diplomat::opaque]
+    pub struct WheelRotations(pub ironrdp::input::WheelRotations);
+
+    impl WheelRotations {
+        pub fn new(is_vertical: bool, rotation_units: i16) -> Box<WheelRotations> {
+            Box::new(WheelRotations(ironrdp::input::WheelRotations {
+                is_vertical,
+                rotation_units,
+            }))
+        }
+
+        pub fn as_operation(&self) -> Box<Operation> {
+            Box::new(Operation(ironrdp::input::Operation::WheelRotations(self.0)))
+        }
+    }
+
+    #[diplomat::opaque]
+    pub struct Scancode(pub ironrdp::input::Scancode);
+
+    impl Scancode {
+        pub fn from_u8(extended: bool, code: u8) -> Box<Scancode> {
+            Box::new(Scancode(ironrdp::input::Scancode::from_u8(extended, code)))
+        }
+
+        pub fn from_u16(code: u16) -> Box<Scancode> {
+            Box::new(Scancode(ironrdp::input::Scancode::from_u16(code)))
+        }
+
+        pub fn as_operation_key_pressed(&self) -> Box<Operation> {
+            let operation = ironrdp::input::Operation::KeyPressed(self.0);
+            Box::new(Operation(operation))
+        }
+
+        pub fn as_operation_key_released(&self) -> Box<Operation> {
+            let operation = ironrdp::input::Operation::KeyReleased(self.0);
+            Box::new(Operation(operation))
+        }
+    }
+
+    #[diplomat::opaque]
+    pub struct Char(pub char);
+
+    impl Char {
+        pub fn new(c: char) -> Box<Char> {
+            Box::new(Char(c))
+        }
+
+        pub fn as_operation_unicode_key_pressed(&self) -> Box<Operation> {
+            let operation = ironrdp::input::Operation::UnicodeKeyPressed(self.0);
+            Box::new(Operation(operation))
+        }
+
+        pub fn as_operation_unicode_key_released(&self) -> Box<Operation> {
+            let operation = ironrdp::input::Operation::UnicodeKeyReleased(self.0);
+            Box::new(Operation(operation))
+        }
```

Comment:
suggestion: Use `u32` and [`char::from_u32`](https://doc.rust-lang.org/std/primitive.char.html#method.from_u32) instead

Reply:
updated